### ISSUE TITLE
JAX float64 path computes pullbacks and ∂∂̄φ in single precision

### ISSUE THREAD

**Issue by DerekMarshall** [ruehlef/cymetric#5]
With `jax_enable_x64=True`, float64 parameters and inputs, and a complex128
`BASIS`, `FSModel.pullbacks` returns `complex64` and `PhiFSModel.__call__` runs
55 of 542 traced equations at single precision before promoting the output. The
`FutureWarning` from `jax/_src/ops/scatter.py` (complex128 scattered into a
complex64 array at `cymetric/jax/models/fubinistudy.py:445`) is a symptom; JAX
will make it an error.

Reproducer: https://gist.github.com/DerekMarshall/2275d30133d013bf9558c0b2f46c5313
Commit c10ce33072f1a250a3f69e5e6953548403b5055c, jax 0.11.1, numpy 2.5.3, Python 3.12.14.

Output:

```
jax 0.11.1 numpy 2.5.3 x64 True
param dtypes: ['float64'] | BASIS dtypes: ['complex128', 'int64'] | x: float64
pullbacks out dtype: ['complex64']
__call__ out dtype: complex128; single-precision eqns: 55/542
sigma numpy float64      : 0.373938043586542
sigma package complex128 : 0.373938046486578  rel dev 7.755e-09
sigma package complex64  : 0.373938053846359  rel dev 2.744e-08
```

σ(g_FS) with all weights zero, 10 000 points, seed 2021:

| | σ | rel. dev. from float64 |
|---|---|---|
| numpy float64 | 0.373938043586542 | — |
| package, complex128 BASIS | 0.373938046486578 | 7.755e-09 |
| package, complex64 BASIS | 0.373938053846359 | 2.744e-08 |

The complex128 result sits 28% of the way from the float64 value to the
complex64 value.

Literal dtypes on the path:

| location | truncated |
|---|---|
| `jax/models/fubinistudy.py:80` | `self.pi` |
| `jax/models/fubinistudy.py:394` | `cpoint` |
| `jax/models/fubinistudy.py:398,400` | patch masks |
| `jax/models/fubinistudy.py:408` | `pb` container |
| `jax/models/fubinistudy.py:445` | scatter of the complex128 `all_dzdz` into it |
| `jax/models/fubinistudy.py:462` | `_indices_to_mask` |
| `jax/models/fubinistudy.py:599-602` | FS metric: `point_square`, `eye`, `outer` |
| `jax/models/fubinistudy.py:648` | `_get_patch_coordinates` |
| `jax/models/models.py:380` | `dd_phi_c` |
| `jax/models/models.py:404` | transition-loss coordinates |

`models.py:383` then contracts the truncated `dd_phi_c` with the pullbacks, and
`models.py:425,428,429` carry the complex64 from line 404 into the network input
for the transition loss. Those are consequences of the casts above, not separate
literals.

A complex128 output is not evidence of a double-precision computation: the last
operation promotes, the arithmetic behind it does not. `pullbacks` does not
promote at all.

With trained weights the effect is larger: `models.py:380` truncates the entire
learned correction, which is zero in the reproducer above.

Related: `FSModel.__init__` (`fubinistudy.py:55-61`) casts to complex64 only when
a `BASIS` entry is a numpy array. A `BASIS` from
`prepare_basis(..., dtype=jnp.complex128)` holds jax arrays and skips the cast,
so precision depends on the container type of the argument.

Suggested fix: one complex dtype attribute on `FSModel`, taken from `BASIS`, used
at the sites above in place of the literals. PR on request.

**Comment by ruehlef** [comment]
Fixed in 77ad946.

Thanks for the precise report — the dtype table and the jaxpr equation count made this straightforward to track down.

## Causes

There were three, and the main one wasn't visible from outside:

1. `FSModel.__init__` cast **every** numpy array in `BASIS` to `complex64`, so a `complex128` BASIS was truncated on entry, before any of the downstream code ran.
2. 88 hardcoded `jnp.complex64` / `jnp.float32` literals across the backend — including all the sites you listed: `self.pi`, `cpoint`, the patch masks, and the scatter at `fubinistudy.py:445`.
3. `helper.prepare_basis` froze its dtype in a default argument evaluated at *import* time, so enabling x64 afterwards could never take effect.

## Fix

Adds `cymetric/jax/models/dtypes.py`, which derives the working precision from `jax.config.jax_enable_x64`, and routes every site through it.

The helpers are functions rather than module-level constants specifically because of cause 3 — x64 can be enabled after import, and a constant captured at import time would silently miss it. Under `jit` the value is a static Python bool, so there is no runtime cost. A complex BASIS entry now also keeps its own width under x64 instead of being downcast.

## Verification

Re-running your scenario (x64 on, `float64` inputs and parameters, `complex128` BASIS, Fermat quintic):

| | reported | now |
| --- | --- | --- |
| `pullbacks` dtype | `complex64` | `complex128` |
| single-precision eqns in `__call__` | 55/542 | **0/2609** |
| BASIS dtypes | `complex128`, `int64` | `complex128` |

The `complex128`-into-`complex64` scatter `FutureWarning` is gone, since both sides now match. Without x64 the behaviour is unchanged — the JAX/TensorFlow equivalence suite still passes 54/54.

## One caveat

`pointgen._generate_padded_basis` still emits a `complex64` BASIS by default, so double precision end to end also requires casting the basis yourself (as you did) or changing the generator. Whether the generator should follow the x64 switch too is a separate decision.
